flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -0,0 +1,1448 @@
+		}
+		assertEquals(totalSizeIncludingLogicallyRemovedStates * 3, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(0, totalStateSize);
+		assertEquals(totalStateSize, stateMap.size());
+		assertEquals(totalSizeIncludingLogicallyRemovedStates, stateMap.totalSize());
+		assertEquals(totalSizeIncludingLogicallyRemovedStates, stateMap.getLogicallyRemovedNodes().size());
+		verifyState(referenceStates, stateMap);
+
+		verifySnapshotWithoutTransform(
+			expectedSnapshot1, snapshot1, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot1.release();
+
+		// no spaces should be free
+		assertEquals(totalSizeIncludingLogicallyRemovedStates * 3, spaceAllocator.getTotalSpaceNumber());
+		verifyState(referenceStates, stateMap);
+
+		Map<Long, Map<Integer, String>> expectedSnapshot2 = snapshotReferenceStates(referenceStates);
+		CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String> snapshot2 =
+			(CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String>) stateMap.stateSnapshot();
+
+		// all state should be removed physically
+		totalSizeIncludingLogicallyRemovedStates = 0;
+		assertEquals(totalStateSize, stateMap.totalSize());
+		assertEquals(totalSizeIncludingLogicallyRemovedStates, stateMap.totalSize());
+		assertEquals(totalSizeIncludingLogicallyRemovedStates, stateMap.getLogicallyRemovedNodes().size());
+		assertEquals(0, spaceAllocator.getTotalSpaceNumber());
+
+		verifySnapshotWithoutTransform(
+			expectedSnapshot2, snapshot2, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot2.release();
+
+		assertEquals(0, stateMap.size());
+		assertEquals(0, stateMap.totalSize());
+		assertEquals(0, spaceAllocator.getTotalSpaceNumber());
+		assertFalse(stateMap.iterator().hasNext());
+
+		stateMap.close();
+	}
+
+	/**
+	 * Tests that snapshots prune useless values.
+	 */
+	@SuppressWarnings("unchecked")
+	@Test
+	public void testSnapshotPruneValues() throws IOException {
+		TypeSerializer<Integer> keySerializer = IntSerializer.INSTANCE;
+		TypeSerializer<Long> namespaceSerializer = LongSerializer.INSTANCE;
+		TypeSerializer<String> stateSerializer = StringSerializer.INSTANCE;
+		// set logicalRemovedKeysRatio to 0 so that all logically removed states will be deleted when snapshot
+		CopyOnWriteSkipListStateMap<Integer, Long, String> stateMap = new CopyOnWriteSkipListStateMap<>(keySerializer,
+			namespaceSerializer,
+			stateSerializer,
+			spaceAllocator);
+
+		// firstly build value chain and snapshots as follows
+		//  ------      ------      ------      ------
+		// |  v3  | -> |  v2  | -> |  v1  | -> |  v0  |
+		//  ------      ------      ------      ------
+		//    |            |           |          |
+		// snapshot4   snapshot3   snapshot2   snapshot1
+		// snapshot5
+		// snapshot6
+
+		List<String> referenceValues = new ArrayList<>();
+
+		// build v0
+		stateMap.put(1, 1L, "0");
+		referenceValues.add(0, "0");
+		// get the pointer to the node
+		long node = stateMap.getLevelIndexHeader().getNextNode(0);
```

Comment:
Isn't this an internal implementation detail? I would like to avoid access to internal state as much as possible.

Reply:
Yes it's an internal implementation detail, but the logic of pruning out-of-date versions is an optimization not accessible from any public API, so we have to touch the internal